src/server/src/main.rs

```diff
@@ -1,23 +1,215 @@
+    email: Option<String>,
+    age: Option<u32>,
+    gender: Option<String>,
 }
+// sessionId -> child process reference
+type SessionMap = Arc<Mutex<HashMap<String, ChildRef>>>;
+// username -> User
+type UserMap = Arc<Mutex<HashMap<String, User>>>;
 
-#[web::get("/{name}")]
-async fn hello(name: web::types::Path<String>) -> impl web::Responder {
-    web::HttpResponse::Ok()
-        .content_type("text/plain")
-        //THIS IS FOR THE DEMO, DO NOT DO THIS IN PRODUCTION GET RID OF THIS SOON
-        .set_header(ntex::http::header::ACCESS_CONTROL_ALLOW_ORIGIN, "*")
-        .body(format!("Hello {}!", &name))
+fn spawn_user_session(username: String) -> ChildRef 
+{
+    Bastion::children(|children| 
+        {
+        children.with_exec(move |ctx| 
+            {
+            let username_clone = username.clone();
+            async move 
+            {
+                loop 
+                {
+                    let msg = ctx.recv().await?;
+                    if let Ok(task) = msg.downcast::<String>() 
+                    {
+                        println!("Session for {} doing: {}", username_clone, task);
+                    }
+                    ctx.reply(format!("done {}", username_clone)).unwrap();
+                }
+                Ok(())
+            }
+        })
+    })
+    .expect("Failed to spawn user session")
+}
+
+async fn create_user_handler(
+    body: web::types::Json<User>,
+    path: web::types::Path<(String,)>,
+    user_map: web::types::Data<UserMap>,) -> impl Responder 
+{
+    let new_user = body.into_inner();
+    let (username,) = path.into_inner();
+    let mut map = user_map.lock().unwrap();
+
+    if map.contains_key(&username) {
+        return HttpResponse::build(418).body("User already exists.");
+    }
+
+    map.insert(username.clone(), new_user);
+    HttpResponse::Ok().body("User successfully created.")
+}
+
+async fn get_user_handler(
+    req: HttpRequest,
+    path: web::types::Path<(String,)>,
+    user_map: web::types::Data<UserMap>,
+    session_map: web::types::Data<SessionMap>,) -> impl Responder 
+{
+    let (username,) = path.into_inner();
+    let session_id = req.headers().get("SessionID").and_then(|val| val.to_str().ok());
+    let user_map = user_map.lock().unwrap();
+
+    if let Some(user) = user_map.get(&username) 
+    {
+        if let Some(session_id) = session_id
+        {
+            let session_map = session_map.lock().unwrap();
+            if let Some(sess_user) = session_map.get(session_id) 
+            {
+                // if current session matches. then return full data
+                println!("Session ID matched for user {}", username);
+                return HttpResponse::Ok().json(user);
+            } 
+        }
+        // else return public data 
+        let public_data = User 
+        {
+            username: user.username.clone(),
+            password: String::from("***"), 
+        };
+        HttpResponse::Ok().json(public_data)
```

Comment:
there shouldn't be a password field at all in the returned JSON. 
also the other optionals have been neglected. (`age` and `gender` should probably be public too, or maybe we have the user configure their privacy settings)

Reply:
so we have to think a little bit harder on how to generalize this, so that we don't have the same issue as we have in the current implementation of the `PUT` call. 
maybe we just serialize the structure to `JSON`, delete the fields declared as private (either hard-coded or user specified), and return that.